### ISSUE TITLE
Charm workloads may not receive a SIGTERM during charm upgrade

### ISSUE THREAD

**Issue by balbirthomas** [canonical/pebble#122]
## Issue
 
I find that in some cases a charm workload may not get a SIGTERM signal when it is shut down prior to being upgraded.

I ran into this issue in trying to develop [this integration test](https://github.com/canonical/alertmanager-k8s-operator/blob/itest-storage/tests/integration/test_persistence.py) for the Alertmanger k8s charm as implemented in this [git branch](https://github.com/canonical/alertmanager-k8s-operator/tree/itest-storage). The purpose of the integration test is to check persistence of [silences](https://prometheus.io/docs/alerting/latest/alertmanager/#silences) across charm upgrade. I have checked using just the Alertmanager [standalone binary](https://github.com/prometheus/alertmanager/releases/tag/v0.23.0) (version 0.23 same as charm's rock) that Alertmanager does indeed flush silences stored in memory to disk whenever it receives either a SIGTERM or a SIGINT signal. I have also checked that the Alertmanager charm does get a bound persistent volume claim (used to store silences under the path `/alertmanger`) which is the same across charm upgrade and that this volume is empty before and after charm upgrade even when a silence has been set and is visible in the Alertmanager dashboard. On manually sending either a SIGTERM or a SIGINT (using juju ssh or kubectl exec), to the Alertmanger charm workload prior to doing a charm upgrade, the silences do get flushed to the persistent volume and are visible across charm upgrade. The integration test also does this using a Juju action to send a SIGINT. Please note the current edge release of Alertmanager does not have this action which is required by the test. However if the Alertmanager charm from the [linked branch](https://github.com/canonical/alertmanager-k8s-operator/tree/itest-storage) (with the action) is deployed, the action invoked and the charm upgraded, silences do persist. As I understand Pebble sends a SIGTERM then a SIGKILL to a workload process when stopping it. Hence I would have expected Alertmanger charm workload will persist silences to disk anytime it is stopped prior to upgrade, without the need to send this signal using a Juju action for instance.

## Impacts

If the charm workload (in this case Alertmanager) does not get a trappable signal like SIGTERM then this may lead to potential data loss since the workload is unable to flush in memory structures to disk and gracefully exit. A workaround may be to provide the charm with a Juju action to issue such a signal prior to charm upgrade. However any such workaround is prone to human error on failing to notice what needs to be done prior to upgrade.

**Comment by rwcarlsen** [comment]
I wonder if this has to do with the mechanics of how juju/k8s handle the upgrade sequence; pebble may not be the culprit.

**Comment by jnsgruk** [comment]
Yeh, I think this may be fixed when @hpidcock and co. finish the finaliser work.

**Comment by benhoyt** [comment]
Copying from my Mattermost message:

Pebble was updated to stop (i.e., send SIGTERM to) its processes when it receives SIGTERM/SIGQUIT: https://github.com/canonical/pebble/pull/114

That was merged into Pebble on 21 June and subsequently merged into Juju (2.9 branch). HOWEVER, the latest release of Juju 2.9, which is 2.9.32, which was released on 13 June (https://github.com/juju/juju/tags), does not include this update.

So there may be upgrade-event issues here too, but the first problem is that this Pebble change needs to be included in a version of Juju near you.

**Comment by benhoyt** [comment]
@jameinel said: "We are definitely doing a 2.9.33 (and almost definitely a future 2.9.34+). 2.9.33 we're likely to get out just before 3.0.0"

**Comment by hpidcock** [comment]
This is likely related to https://bugs.launchpad.net/juju/+bug/1951415

**Comment by benhoyt** [comment]
This was a Juju issue and should be fixed by the bug linked above. Closing this.
